src/session/content/chat_action_bar.rs

```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
+                                !data.is_channel()
+                            } else if let ChatMemberStatus::Restricted(data) = data.status().0 {
+                                !data.permissions.can_send_messages
+                            } else {
+                                !chat.permissions().0.can_send_messages
+                            }
+                        } else if let ChatType::BasicGroup(data) = chat.type_() {
+                            if let ChatMemberStatus::Restricted(data) = data.status().0 {
+                                !data.permissions.can_send_messages
+                            } else {
+                                !chat.permissions().0.can_send_messages
+                            }
+                        } else {
+                            chat.permissions().0.can_send_messages
+                        }
+                    } else {
+                        match chat.type_() {
+                            ChatType::Supergroup(data) => match data.status().0 {
+                                ChatMemberStatus::Restricted(data) => {
+                                    !data.permissions.can_send_messages
+                                }
+                                _ => !chat.permissions().0.can_send_messages,
+                            },
+                            ChatType::BasicGroup(data) => match data.status().0 {
+                                ChatMemberStatus::Restricted(data) => {
+                                    !data.permissions.can_send_messages
+                                }
+                                _ => !chat.permissions().0.can_send_messages,
+                            },
+                            _ => !chat.permissions().0.can_send_messages,
+                        }
+                    }
+                })
+                .upcast()
+                .bind(&*imp.restriction_label, "visible", Some(chat));
+
+            bindings.push(visible_binding);
+
+            // Handle restricion_label caption
+            if let ChatType::Supergroup(data) = chat.type_() {
+                let restriction_binding = data.restriction_expression(permissions_expression).bind(
+                    &*imp.restriction_label,
+                    "label",
+                    Some(data),
+                );
+                bindings.push(restriction_binding)
+            } else if let ChatType::BasicGroup(data) = chat.type_() {
+                let restriction_binding = data.restriction_expression(permissions_expression).bind(
+                    &*imp.restriction_label,
+                    "label",
+                    Some(data),
+                );
+                bindings.push(restriction_binding)
+            }
+            // Neither of Private or Secret chats are using restricted_label so we don't need else statement
         }
 
+        // Connect scrolled_window (message area) visibility with send_message_button visibility
+        self.imp()
```

Comment:
I guess you wanted to do that in the `constructed()` function. But anyway I don't think this is a good idea. I think it would be better to connect the previous `visible_expression` to both the `scrolled_window` and the `send_message_button`.